### PR TITLE
Introduce CreateConferenceForm form object

### PR DESCRIPTION
**Pull request by adamwathan**
Pulled out the logic from the `ConferenceController@store` action into a form object. A little different than going the FormRequest route but I think these sorts of objects are nicer to test and easier to follow. You don't have to know anything about how the framework magically instantiates a certain request object or calls certain validation methods on it, etc.

Feel free to ignore if you want to use a FormRequest instead, it was fun either way :)

### CODE REVIEW

File: tests/CreateConferenceFormTest.php [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        $input = [
+            'title' => 'AwesomeConf 2015',
+            'description' => 'The best conference in the world!',
+            'url' => 'http://example.com',
+        ];
+
+        $form = CreateConferenceForm::fillOut($input, Factory::create('user'));
+        $conference = $form->complete();
+
+        $this->assertInstanceOf(Conference::class, $conference);
```
Comment (mattstauffer) [inline tests/CreateConferenceFormTest.php]: I think this doesnt' work on 5.4, but I don't think I care if this works on 5.4, so I'll update our travis settings for that later. :+1:

Comment (adamwathan) [inline tests/CreateConferenceFormTest.php]: Stupid PHP 5.4

**Comment by adamwathan** [comment]
Also, I'm personally using this project as a bit of an exercise in good naming and descriptive code, so please question everything :)

I know right now I'm trying to think of a better name for the "CreateConferenceForm" for example. Gut reaction was to use "create" because I'm thinking like a developer, but the user isn't really creating a conference. Maybe submitting? They are trying to get the conference added, maybe AddConference? Seems not specific enough though.

Same with `$form->complete()`. I think it's probably more like `process` but that word is a bit of a smell like `Manager` :) But if I think of like a passport office analogy and I'm filling out a form and submitting it, I think they *would* use the word process, so maybe it's ok.

Conflicted on the `ValidationException` approach as well. I used to be totally fine with it, justifying it as a "Tell, don't ask" sort of thing, where if I'm telling the form "process" it should just blow up if anything goes wrong and let me know, rather than interrogating the form in advance and asking "are you valid? Ok you are, now process". But either way the controller knows about the potential for validation to fail, so it's not really hiding any concerns. Knowing to catch a validation exception is basically the same as asking it if it's valid.

The other option is the controller passing itself to the form when processing it as a context, and the form can call the right method on the controller should validation fail. This is probably the most "pure" way of doing it, but it might be unnecessarily obfuscated vs. the more procedural approach.

File: app/Services/CreateConferenceForm.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    private $input;
+    private $user;
+
+    private function __construct($input, $user)
+    {
+        $this->input = $this->removeEmptyFields($input);
+        $this->user = $user;
+    }
+
+    private function removeEmptyFields($input)
```
Comment (mattstauffer) [inline app/Services/CreateConferenceForm.php]: Does this handle intentionally nulling out a date? Guessing we need to add a test for that? @adamwathan

File: tests/CreateConferenceFormTest.php [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        $this->assertEquals(new DateTime('2015-02-01'), $conference->starts_at);
+        $this->assertEquals(new DateTime('2015-02-04'), $conference->ends_at);
+        $this->assertEquals(new DateTime('2015-01-15'), $conference->cfp_starts_at);
+        $this->assertEquals(new DateTime('2015-01-18'), $conference->cfp_ends_at);
+    }
+
+    /**
+     * @test
+     */
+    public function empty_dates_are_treated_as_null()
```
Comment (mattstauffer) [inline tests/CreateConferenceFormTest.php]: I think if we can just duplicate this test with one that saves it Non-null, and then re-submits it with null dates, and make sure it overrides the old ones.

**Comment by mattstauffer** [comment]
This looks great. Let's drop PHP 5.4 from testing and then add the nullable date testing and then we're good. Either of us can do this.

File: app/Services/CreateConferenceForm.php (hunk 2 above)
Comment (adamwathan) [inline app/Services/CreateConferenceForm.php]: Yeah, that will strip out anything that comes in as an empty string. As I'm pretty sure you noticed, submitting a `starts_at` as empty for example will actually trigger MySQL into storing `0000-00-00` even though the default value for the column is `NULL`.

This is all only for brand new conferences though, not updating existing ones, all that is still in the controller in the other action.

File: tests/CreateConferenceFormTest.php (hunk 3 above)
Comment (adamwathan) [inline tests/CreateConferenceFormTest.php]: Yeah like I mentioned above this is just for adding new conferences, I think what you're talking about matters for updating a conference but we still have that in the controller at the moment, nothing has been extracted there. Let's get that out next and write some tests around it then :)

Comment (mattstauffer) [inline tests/CreateConferenceFormTest.php]: Aha! duh! :)

**Comment by mattstauffer** [comment]
Thanks @adamwathan!

**Pull request merged**